src/ui/public/index_patterns/_index_pattern.js

```diff
@@ -13,318 +13,334 @@ import IndexPatternsFieldListProvider from 'ui/index_patterns/_field_list';
+    indexPattern._docSource
+      .onUpdate()
+      .then(response => updateFromElasticSearch(indexPattern, response), notify.fatal);
+  }
 
-    let docSource = new DocSource();
+  class IndexPattern {
+    constructor(id) {
+      this._setId(id);
+      this._docSource = new DocSource();
+      this.metaFields = config.get('metaFields');
+      this.getComputedFields = getComputedFields.bind(this);
 
-    self.init = function () {
-      // tell the docSource where to find the doc
-      docSource
-      .index(kbnIndex)
-      .type(type)
-      .id(self.id);
+      this.flattenHit = flattenHit(this);
+      this.formatHit = formatHit(this, fieldformats.getDefaultInstance('string'));
+      this.formatField = this.formatHit.formatField;
+
+      this.routes = IndexPattern.routes;
+    }
+
+    _setId(id) {
+      this.id = id;
+      return id;
+    }
+
+    _watch() {
+      if (this._watching) {
+        return;
+      }
+      this._watching = true;
 
-      // listen for config changes and update field list
       config.watchAll(() => {
-        if (self.fields) {
-          // re-init fields when config changes, but only if we already had fields
-          initFields();
+        if (this._fieldsInitialized) {
+          this._initFields(); // re-init fields when config changes, but only if we already had fields
         }
       });
+    }
 
-      return mappingSetup.isDefined(type)
-      .then(function (defined) {
-        // create mapping for this type if one does not exist
-        if (defined) return true;
-        return mappingSetup.setup(type, mapping);
-      })
-      .then(function () {
-        // If there is no id, then there is no document to fetch from elasticsearch
-        if (!self.id) return;
-
-        // fetch the object from ES
-        return docSource.fetch()
-        .then(function applyESResp(resp) {
-          if (!resp.found) throw new errors.SavedObjectNotFound(type, self.id);
-
-          // deserialize any json fields
-          _.forOwn(mapping, function ittr(fieldMapping, name) {
-            if (fieldMapping._deserialize) {
-              resp._source[name] = fieldMapping._deserialize(resp._source[name], resp, name, fieldMapping);
-            }
-          });
-
-          // Give obj all of the values in _source
-          _.assign(self, resp._source);
-
-          self._indexFields();
-
-          // Any time obj is updated, re-call applyESResp
-          docSource.onUpdate().then(applyESResp, notify.fatal);
-        });
-      })
-      .then(function () {
-        // return our obj as the result of init()
-        return self;
-      });
-    };
-
-    function initFields(fields) {
-      self.fields = new FieldList(self, fields || self.fields || []);
+    init() {
+      this._docSource
+        .index(kbnIndex)
+        .type(type)
+        .id(this.id);
+
+      this._watch();
+
+      return mappingSetup
+        .isDefined(type)
+        .then(defined => {
```

Comment:
These lines should be indented at the same level as `return` https://github.com/elastic/kibana/blob/master/STYLEGUIDE.md#chaining-operations